src/number_theory/multiplicity.lean

```diff
@@ -0,0 +1,412 @@
+  have h : (multiplicity _ _).dom := finite_nat_iff.mpr ⟨hp.ne_one, n.succ_pos⟩,
+  rcases eq_coe_iff.mp (part_enat.coe_get h).symm with ⟨⟨k, hk⟩, hpn⟩,
+  conv_lhs { rw [hk, pow_mul, pow_mul] },
+  rw nat.prime_iff_prime_int at hp,
+  rw [pow_sub_pow_of_prime hp, pow_prime_pow_sub_pow_prime_pow hp hp1 hxy hx, part_enat.coe_get],
+  { rw ←geom_sum₂_mul,
+    exact dvd_mul_of_dvd_right hxy _ },
+  { exact λ h, hx (hp.dvd_of_dvd_pow h) },
+  { rw int.coe_nat_dvd,
+    rintro ⟨c, rfl⟩,
+    refine hpn ⟨c, _⟩,
+    rwa [pow_succ', mul_assoc] }
+end
+
+
+lemma int.pow_add_pow {x y : ℤ} (hxy : ↑p ∣ x + y) (hx : ¬↑p ∣ x) {n : ℕ} (hn : odd n) :
+  multiplicity ↑p (x ^ n + y ^ n) = multiplicity ↑p (x + y) + multiplicity p n :=
+begin
+  rw ←sub_neg_eq_add at hxy,
+  rw [←sub_neg_eq_add, ←sub_neg_eq_add, ←odd.neg_pow hn],
+  exact int.pow_sub_pow hp hp1 hxy hx n
+end
+
+lemma nat.pow_sub_pow {x y : ℕ} (hxy : p ∣ x - y) (hx : ¬p ∣ x) (n : ℕ) :
+  multiplicity p (x ^ n - y ^ n) = multiplicity p (x - y) + multiplicity p n :=
+begin
+  obtain hyx | hyx := le_total y x,
+ { iterate 2 { rw ← int.coe_nat_multiplicity },
+    rw int.coe_nat_sub (nat.pow_le_pow_of_le_left hyx n),
+    rw ← int.coe_nat_dvd at hxy hx,
+    push_cast at *,
+    exact int.pow_sub_pow hp hp1 hxy hx n },
+  { simp only [nat.sub_eq_zero_iff_le.mpr hyx,
+      nat.sub_eq_zero_iff_le.mpr (nat.pow_le_pow_of_le_left hyx n), multiplicity.zero,
+      part_enat.top_add] }
+end
+
+lemma nat.pow_add_pow {x y : ℕ} (hxy : p ∣ x + y) (hx : ¬p ∣ x) {n : ℕ} (hn : odd n) :
+  multiplicity p (x ^ n + y ^ n) = multiplicity p (x + y) + multiplicity p n :=
+begin
+  iterate 2 { rw [←int.coe_nat_multiplicity] },
+  rw ←int.coe_nat_dvd at hxy hx,
+  push_cast at *,
+  exact int.pow_add_pow hp hp1 hxy hx hn
+end
+
+end lifting_the_exponent
+end multiplicity
+end comm_ring
+
+lemma  pow_two_pow_sub_pow_two_pow [comm_ring R] {x y : R} (n : ℕ) :
+  x ^ (2 ^ n) - y ^ (2 ^ n) = (∏ i in finset.range n, (x ^ (2 ^ i) + y ^ (2 ^ i))) * (x - y) :=
+begin
+  induction n with d hd,
+  { simp only [pow_zero, pow_one, finset.range_zero, finset.prod_empty, one_mul] },
+  { suffices : x ^ 2 ^ d.succ - y ^ 2 ^ d.succ = (x ^ 2 ^ d + y ^ 2 ^ d) * (x ^ 2 ^ d - y ^ 2 ^ d),
+    { rw [this, hd, finset.prod_range_succ, ← mul_assoc, mul_comm (x ^ 2 ^ d + y ^ 2 ^ d)] },
+    { ring_exp_eq } }
+end
+
+lemma int.sq_mod_four_eq_one_of_odd {x : ℤ} : odd x → x ^ 2 % 4 = 1 :=
+begin
+  intro hx,
+  -- Replace `x : ℤ` with `y : zmod 4`
+  replace hx : x % (2 : ℕ) = 1 % (2 : ℕ), { rw int.odd_iff at hx, norm_num [hx] },
+  calc x^2 % (4 : ℕ)
+      = 1 % (4 : ℕ) : _
+  ... = 1 : by norm_num,
+  rw ← zmod.int_coe_eq_int_coe_iff' at hx ⊢,
+  push_cast,
+  rw [← map_int_cast (zmod.cast_hom (show 2 ∣ 4, by norm_num) (zmod 2)) x] at hx,
+  set y : zmod 4 := x,
+  -- Now we can just consider each of the 4 possible values for y
+  fin_cases y using hy;
+    rw hy at ⊢ hx; revert hx; dec_trivial
+end
+
+namespace multiplicity
```

Comment:
As Anne suggested above, it's probably more clear if you don't close/reopen the `multiplicity` namespace, but just use `_root_` when needed.